server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java

```diff
@@ -89,12 +123,35 @@ public Decision canRebalance(RoutingAllocation allocation) {
                 clusterConcurrentRebalance
             );
         }
+        if (frozenRelocationUnlimited == false && relocatingFrozenShards >= clusterConcurrentFrozenRebalance) {
+            return allocation.decision(
+                Decision.THROTTLE,
+                NAME,
+                "reached the limit of concurrently rebalancing frozen shards [%d], cluster setting [%s=%d]",
+                relocatingFrozenShards,
+                CLUSTER_ROUTING_ALLOCATION_CLUSTER_CONCURRENT_FROZEN_REBALANCE_SETTING.getKey(),
+                clusterConcurrentFrozenRebalance
+            );
+        }
         return allocation.decision(
             Decision.YES,
             NAME,
-            "below threshold [%d] for concurrent rebalances, current rebalance shard count [%d]",
+            "below threshold [%d] for concurrent rebalances, current rebalance shard count [%d], "
+                + "and threshold [%d] for concurrent frozen rebalances, current frozen rebalance shard count [%d]",
             clusterConcurrentRebalance,
-            relocatingShards
+            relocatingShards,
+            clusterConcurrentFrozenRebalance,
+            relocatingFrozenShards
         );
     }
+
+    private int countRelocatingFrozenShards(RoutingAllocation allocation) {
+        int frozenRelocations = 0;
+        for (var routingNode : allocation.routingNodes()) {
+            if (routingNode.node().isDedicatedFrozenNode()) {
+                frozenRelocations += routingNode.shardCountsWithState(ShardRoutingState.RELOCATING);
+            }
+        }
+        return frozenRelocations;
+    }
```

Comment:
`canRebalance(shard, allocation)` can be called once for a very large number of shards, so it might be worth maintaining this count in `RoutingNodes` a-la `relocatingShardCount` instead of recalculating it every time